src/UserSpaceInstrumentation/Attach.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "UserSpaceInstrumentation/Attach.h"
+
+#include <errno.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <chrono>
+#include <memory>
+#include <set>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "OrbitBase/GetProcessIds.h"
+#include "OrbitBase/Logging.h"
+#include "absl/strings/str_format.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::GetTidsOfProcess;
+
+namespace {
+
+bool ThreadExistsInProcess(pid_t pid, pid_t tid) {
+  const auto process_tids = orbit_base::GetTidsOfProcess(pid);
+  return std::find(process_tids.begin(), process_tids.end(), tid) != process_tids.end();
+}
+
+// Attaches to the thread using ptrace.
+// Returns true if the thread was halted and false if the thread did not exist anymore.
+[[nodiscard]] ErrorMessageOr<bool> AttachAndStopThread(pid_t pid, pid_t tid) {
+  if (ptrace(PTRACE_ATTACH, tid, 0, 0) < 0) {
+    if (ThreadExistsInProcess(pid, tid)) {
+      return ErrorMessage(
+          absl::StrFormat("Can not attach to tid \"%d\": \"%s\"", tid, strerror(errno)));
+    } else {
+      return false;
+    }
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int timeout = 1000;
+  while (true) {
+    const int result_waitpid = waitpid(tid, nullptr, WNOHANG);
+    if (result_waitpid == -1) {
+      if (ThreadExistsInProcess(pid, tid)) {
+        return ErrorMessage(
+            absl::StrFormat("Can not attach to tid \"%d\": \"%s\"", tid, strerror(errno)));
+      } else {
+        return false;
+      }
+    }
+    if (result_waitpid > 0) {
+      // waitpid returns greater zero occationally for terminated threads. We catch this here so we
```

Comment:
This comment confused me as it seems this whole case is for catching the rare case for > 0 is returned for terminated threads. But if I understand correctly, > 0 is the case we actually want to wait for. So the comment only really refers to the additional check if the thread exists in the process. 

Maybe start the comment with something like:
" This is the success case we wait for and where we want to return true. However, sometimes waitipid returns greater zero also for terminated threads..."

Reply:
I understand way better now what happens. The thread exits before it gets descheduled and that is what causes waitpid to return. The cases can be disambiguated by the status value. 
Adjusted comment accordingly.